src/index/mod.rs

```diff
@@ -0,0 +1,47 @@
+pub mod options;
+
+use crate::bson::Document;
+
+use self::options::*;
+use serde::{Deserialize, Serialize};
+
+use typed_builder::TypedBuilder;
+
+/// Specifies the fields and options for an index. For more information, see the [documentation](https://docs.mongodb.com/manual/indexes/).
+#[derive(Clone, Debug, Default, Deserialize, TypedBuilder, Serialize)]
+#[builder(field_defaults(default, setter(into)))]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub struct IndexModel {
+    /// Specifies the index’s fields. For each field, specify a key-value pair in which the key is
+    /// the name of the field to index and the value is index type.
+    #[serde(rename = "key")]
+    keys: Document,
+
+    /// The options for the index.
+    #[serde(flatten)]
```

Comment:
I'm curious why `IndexOptions` are split out as their own struct; is it ever useful to be dealing with the options distinct from the index keys?

Reply:
I think it's just a convenience - `keys` is the only part of the command which the driver strictly requires. To be honest with you I don't love it because any index returned from the server will include the `name` option, so `IndexOptions` will never be `None` in practice (except if the user is creating an index with no name specified).

But, the `keys` / `index_options` pattern is what the specification requires so I'd argue for keeping it at the end of the day